File: scripts/docs-new.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import re
import sys
from pathlib import Path


ALLOWED_DOC_TYPES = {
    "overview",
    "architecture",
    "guide",
    "reference",
    "troubleshooting",
    "decision",
    "runbook",
}


def title_from_slug(slug: str) -> str:
    name = Path(slug).stem if slug.endswith(".md") else Path(slug).name
    return " ".join(part.capitalize() for part in re.split(r"[-_]+", name) if part)


def page_key(slug: str) -> str:
    key = slug.strip().strip("/")
    if key.startswith("docs/"):
        key = key[len("docs/") :]
    if key.endswith(".md"):
        key = key[:-3]
    return key


def page_path(docs_dir: Path, slug: str) -> Path:
    key = page_key(slug)
    path = docs_dir / f"{key}.md"
    resolved = path.resolve()
    if not str(resolved).startswith(str(docs_dir.resolve())):
        raise ValueError("page slug must stay under docs/")
    return path


def frontmatter(title: str, summary: str, doc_type: str, read_when: list[str]) -> str:
    hints = "\n".join(f"  - {hint}" for hint in read_when)
    return f"""---
doc-schema-version: 1
title: "{title}"
summary: "{summary}"
read_when:
{hints}
doc_type: "{doc_type}"
---

# {title}

Add the useful project-specific content here.
"""


def collect_groups(value: object) -> list[dict[str, object]]:
    groups: list[dict[str, object]] = []
    if isinstance(value, list):
        for item in value:
            groups.extend(collect_groups(item))
    elif isinstance(value, dict):
        if isinstance(value.get("group"), str) and isinstance(value.get("pages"), list):
            groups.append(value)
        for item in value.values():
            groups.extend(collect_groups(item))
    return groups


def add_to_nav(docs_json_path: Path, key: str, group_name: str) -> None:
    if not docs_json_path.exists():
        return

    config = json.loads(docs_json_path.read_text(encoding="utf-8"))
    navigation = config.setdefault("navigation", [])
    if not isinstance(navigation, list):
        raise ValueError("docs/docs.json navigation must be a list")

    groups = collect_groups(navigation)
    target_group = next((group for group in groups if group.get("group") == group_name), None)
    if target_group is None:
        target_group = {"group": group_name, "pages": []}
        navigation.append(target_group)

    pages = target_group.setdefault("pages", [])
    if not isinstance(pages, list):
        raise ValueError(f"docs/docs.json group `{group_name}` pages must be a list")
    if key not in pages:
        pages.append(key)

    docs_json_path.write_text(json.dumps(config, indent=2) + "\n", encoding="utf-8")


def main() -> int:
    parser = argparse.ArgumentParser(description="Create a docs page with standard frontmatter.")
    parser.add_argument("slug", help="Docs slug, for example `guides/setup` or `reference/cli`.")
    parser.add_argument("--title", help="Display title. Defaults to a title derived from the slug.")
    parser.add_argument("--summary", help="One-sentence page summary.")
    parser.add_argument("--type", choices=sorted(ALLOWED_DOC_TYPES), default="guide", dest="doc_type")
    parser.add_argument("--read-when", action="append", default=[], help="Routing hint. May be repeated.")
    parser.add_argument("--nav-group", help="Append the page to this docs/docs.json navigation group.")
    parser.add_argument("--force", action="store_true", help="Overwrite an existing docs page.")
    args = parser.parse_args()

    docs_dir = Path("docs")
    docs_dir.mkdir(parents=True, exist_ok=True)

    key = page_key(args.slug)
    title = args.title or title_from_slug(key)
    summary = args.summary or f"Documentation page for {title}."
    read_when = args.read_when or [f"Working on {title}"]
    target = page_path(docs_dir, key)

    if target.exists() and not args.force:
        print(f"docs:new: {target} already exists; pass --force to overwrite", file=sys.stderr)
        return 1

    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_text(frontmatter(title, summary, args.doc_type, read_when), encoding="utf-8")

    if args.nav_group:
        add_to_nav(docs_dir / "docs.json", key, args.nav_group)

    print(f"docs:new: wrote {target}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/docs-new.mjs
#!/usr/bin/env node
// Create a docs page with the standard frontmatter, optionally wiring it into docs.json nav.
//
//   pnpm run docs:new -- guides/setup --title "Setup" --read-when "Setting the repo up"

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs'
import path from 'node:path'
import { ALLOWED_DOC_TYPES } from './docs-frontmatter.mjs'
import { ROOT, parseCliOrExit } from './script-utils.mjs'

/**
 * `guides/my-page.md`, `/docs/guides/my-page`, … → `guides/my-page`.
 * @param {string} slug
 * @returns {string}
 */
function pageKey(slug) {
	let key = slug.trim().replace(/^\/+|\/+$/g, '')
	if (key.startsWith('docs/')) key = key.slice('docs/'.length)
	return key.replace(/\.md$/, '')
}

/**
 * `my-new-page` → `My New Page`.
 * @param {string} slug
 * @returns {string}
 */
function titleFromSlug(slug) {
	return path
		.basename(slug)
		.split(/[-_]+/)
		.filter(Boolean)
		.map((part) => part.charAt(0).toUpperCase() + part.slice(1))
		.join(' ')
}

/**
 * Every `{ group, pages }` object anywhere in a navigation tree.
 * @param {unknown} value
 * @returns {NavGroup[]}
 */
function collectGroups(value) {
	/** @type {NavGroup[]} */
	const groups = []
	if (Array.isArray(value)) {
		for (const item of value) groups.push(...collectGroups(item))
	} else if (value && typeof value === 'object') {
		const record = /** @type {Record<string, unknown>} */ (value)
		if (typeof record.group === 'string' && Array.isArray(record.pages)) groups.push(/** @type {NavGroup} */ (record))
		for (const item of Object.values(record)) groups.push(...collectGroups(item))
	}
	return groups
}

/**
 * One navigation group in `docs/docs.json`.
 * @typedef {{group: string, pages: string[]}} NavGroup
 */

/**
 * @param {string} docsJsonPath
 * @param {string} key the page key to add
 * @param {string} groupName the group to add it under, created if absent
 * @returns {void}
 */
function addToNav(docsJsonPath, key, groupName) {
	if (!existsSync(docsJsonPath)) return
	const config = JSON.parse(readFileSync(docsJsonPath, 'utf8'))
	config.navigation ??= []
	if (!Array.isArray(config.navigation)) throw new Error('docs/docs.json navigation must be a list')

	let target = collectGroups(config.navigation).find((group) => group.group === groupName)
	if (!target) {
		target = { group: groupName, pages: [] }
		config.navigation.push(target)
	}
	if (!Array.isArray(target.pages)) throw new Error(`docs/docs.json group \`${groupName}\` pages must be a list`)
	if (!target.pages.includes(key)) target.pages.push(key)

	writeFileSync(docsJsonPath, JSON.stringify(config, null, 2) + '\n', 'utf8')
}

const USAGE = `Create a docs page with the standard frontmatter, optionally wiring it into docs.json nav.

  pnpm run docs:new -- guides/setup --title "Setup" --read-when "Setting the repo up"

Options:
  --title <text>       page title (default: derived from the slug)
  --summary <text>     one-line summary for the frontmatter
  --type <type>        one of ${ALLOWED_DOC_TYPES.join(', ')} (default: guide)
  --read-when <hint>   a read_when hint; repeat for more than one
  --nav-group <name>   docs.json nav group to add the page to
  --force              overwrite an existing page
  -h, --help           show this message`

const { values, positionals } = parseCliOrExit(process.argv.slice(2), {
	usage: USAGE,
	allowPositionals: true,
	options: {
		title: { type: 'string' },
		summary: { type: 'string' },
		type: { type: 'string', default: 'guide' },
		'read-when': { type: 'string', multiple: true, default: [] },
		'nav-group': { type: 'string' },
		force: { type: 'boolean', default: false },
	},
})

const slug = positionals[0]
if (!slug) {
	console.error('docs:new: a page slug is required\n')
	console.error(USAGE)
	process.exit(2)
}
if (!ALLOWED_DOC_TYPES.includes(values.type)) {
	console.error(`docs:new: --type must be one of: ${ALLOWED_DOC_TYPES.join(', ')}`)
	process.exit(2)
}

// Resolved from the repo root, not the caller's cwd: `pnpm run` happens to set the cwd to the
// package root, but `node scripts/docs-new.mjs` from a subdirectory would otherwise scaffold a
// second `docs/` tree there.
const docsDir = path.resolve(ROOT, 'docs')
mkdirSync(docsDir, { recursive: true })

const key = pageKey(slug)
const title = values.title || titleFromSlug(key)
const summary = values.summary || `Documentation page for ${title}.`
// `parseCli` types its `values` loosely (see its doc), so the `multiple: true` flag arrives
// without its element type; the annotation is what the template literal below needs.
/** @type {string[]} */
const readWhen = values['read-when'].length > 0 ? values['read-when'] : [`Working on ${title}`]

const target = path.resolve(docsDir, `${key}.md`)
if (!target.startsWith(docsDir + path.sep)) {
	console.error('docs:new: page slug must stay under docs/')
	process.exit(1)
}
if (existsSync(target) && !values.force) {
	console.error(`docs:new: ${target} already exists; pass --force to overwrite`)
	process.exit(1)
}

mkdirSync(path.dirname(target), { recursive: true })
writeFileSync(
	target,
	`---
doc-schema-version: 1
title: "${title}"
summary: "${summary}"
read_when:
${readWhen.map((hint) => `  - ${hint}`).join('\n')}
doc_type: "${values.type}"
---

# ${title}

Add the useful project-specific content here.
`,
	'utf8'
)

if (values['nav-group']) addToNav(path.join(docsDir, 'docs.json'), key, values['nav-group'])

console.log(`docs:new: wrote ${target}`)
